packages/cli/src/lib/bundler/transforms.ts

```diff
@@ -32,6 +32,32 @@ export const transforms = (options: TransformOptions): Transforms => {
 
   const extraTransforms = isDev ? ['react-hot-loader'] : [];
 
+  // Make sure that imported css styles are included before style
+  // elements created by makeStyle to keep a stable insertion
+  // order which results in a stable evaluation order of the css
+  // selectors.
+  //
+  // insert is executed in the browser without transformation,
+  // avoid using new JS features here!
+  function insertElementAfterTitle(element: any) {
```

Comment:
I'm a bit worried this is gonna have different behavior between dev and prod builds due `inject` only applying to async chunks in the latter. Perhaps not for the changes in this PR, but in general.

Do you think there are some other options here? I assume not since we're in the CLI package 😅. Gonna play around with this a bit myself since I wanna understand the issue a bit more, I assume reproducing the issue is simple :p

Reply:
Hmm I actual assumed that it works the same way in prod and dev. I verified it and tested that the insertion point works in both cases. Which difference do you mean exactly?

Reply:
Isn't the only difference between prod and dev that it either uses `<link>` or `<style>`?

Reply:
The `insert` option of the `MiniCssExtractPlugin` only applies to async chunks, https://webpack.js.org/plugins/mini-css-extract-plugin/#insert, while the style-loader applies it for all chunks https://webpack.js.org/loaders/style-loader/#insert.

My worry is that a synchronously imported stylesheet will get higher priority in dev than in prod. You won't see that issue in this repo because we don't import many stylesheets, but it can ofc happen in other places.

Reply:
Tried it out on this branch and it's indeed the case:

```diff
diff --git a/packages/app/public/index.html b/packages/app/public/index.html
index d86f09f0b0..e17487f2ce 100644
--- a/packages/app/public/index.html
+++ b/packages/app/public/index.html
@@ -99,6 +99,12 @@
       });
     </script>
     <% } %>
+
+    <style>
+      body {
+        border: 10px solid #00f;
+      }
+    </style>
   </head>
 
   <body style="margin: 0">
diff --git a/packages/app/src/index.tsx b/packages/app/src/index.tsx
index b15bc4c102..281b520ceb 100644
--- a/packages/app/src/index.tsx
+++ b/packages/app/src/index.tsx
@@ -19,4 +19,6 @@ import React from 'react';
 import ReactDOM from 'react-dom';
 import App from './App';
 
+import './stylish.css';
+
 ReactDOM.render(<App />, document.getElementById('root'));
diff --git a/packages/app/src/stylish.css b/packages/app/src/stylish.css
new file mode 100644
index 0000000000..e2137bd984
--- /dev/null
+++ b/packages/app/src/stylish.css
@@ -0,0 +1,3 @@
+body {
+  border: 10px solid #f00;
+}
```

Dev:

![image](https://user-images.githubusercontent.com/4984472/140981222-fa8e3170-c78e-4efd-91bc-65fc27d0627c.png)

Prod:

![image](https://user-images.githubusercontent.com/4984472/140981255-79b6cb00-ba7f-4524-97f2-88e498826057.png)

Reply:
Wouldn't that speak for an explicit insertion point?

Reply:
Ah , nevermind, got your example. Hmm, doesn't the loader provide any way to control that?

Reply:
I tried your code out, at least there is one good thing: Static CSS is still loaded before dynamic CSS, because the `<link>` elements are inserted quite early. With static I mean all imports and dynamic everything from `makeStyles`.

The whole lazy loading thing makes the order difficult, but if we want to have lazy loading we need to think about it. What guarantees on the order can we give? If I had a wish list, it would be:

1. In every module, I would like to guarantee that the order of imports is preserved:
  ```ts
  import "first.css";
  import "second.css";
  ```
2. In every module, I want that dynamic styles, e.g. created via `makeStyles` have the highest priority.
  ```
  const useStyles = makeStyles({
    root: {
      // Third
    },
  });
  ```

These two rules are working with the current state of the PR. Prepending to `<head>` would make the 1. rule difficult, at least in the naive implementation. I could imaging a third role, but I'm unsure about it.

3. CSS imported in non-async modules come first. This make sure I can import a global theming, e.g. to style material UI via the [Global CSS override](https://mui.com/customization/how-to-customize/#5-global-css-override) approach. I wouldn't make assumptions about the order of imports between all non-async modules, but still support 1. and 2..

The third rule makes it difficult. Sure, you could directly link a CSS file in the HTML, but then you would take advantage of the build process I guess (e.g. minification). I wouldn't make 3. to broad, e.g. guaranteeing any other order between modules, as I think one should prefer to build isolated styles. One could also argue that the *Global CSS override* approach is not supported and one should use other ways of customizations. Like passing overrides into the theme or using `GlobalStyles` (which seems to be targeted on this problem).

I would vote for not supporting rule 3., as it makes it quite difficult, but also still leaves a lot of room for different orders in non-async modules. I think if we decide on something, we should also make sure to document it.

Reply:
Just to try to break this down a little bit, is the main issue right now that dynamically loaded style chunks don't always get placed before `makeStyles` styles in related chunks?

Reply:
Yep 2. is the issue we have in this PR.

Reply:
@Rugvip pointed out that we actually don't have to settle on the order yet. Instead we could make the selectors more specific - This is now part of this PR and the `@backstage/cli` changes got removed. 🎉